### PR TITLE
Add MT-Bench and PR-Bench Support

### PR DESCRIPTION
**Pull request by danmcp**

### CODE REVIEW

File: src/instructlab/eval/mtbench.py [hunk 1]
```diff
@@ -9,48 +12,85 @@ class MT_Bench_Evaluator(Evaluator):
...
             overall_score   MT-Bench score for the overall model evaluation
             qa_pairs        Question and answer pairs from the evaluation
         """
-        overall_score: float = 0.0
-        qa_pairs: list[tuple] = []
-        return overall_score, qa_pairs
+        return gen_judgment.run(
+            self.model_name,
+            judge_model_name=judge_model_name,
+            parallel=40,
```
Comment (alimaredia) [inline src/instructlab/eval/mtbench.py:42]: Did we decide if we want to set this value somewhere or just hardcode 40 somehow? That's a smaller detail and shouldn't hold up the merging of this PR.

File: src/instructlab/eval/gen_api_answer.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+def run(
+    model_name,
+    question_begin=None,
+    question_end=None,
+    force_temperature=None,
+    answer_file=None,
+    num_choices=1,
+    max_tokens=1024,
+    parallel=1,
+    openai_api_base=None,
```
Comment (alimaredia) [inline src/instructlab/eval/gen_api_answer.py:89]: Maybe we rename this arg to `judge_endpoint` just so that the user is not confused with the mention of `openai` in the arugment

File: src/instructlab/eval/gen_api_answer.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+):
+    if openai_api_base is not None:
+        openai.api_base = openai_api_base
+
+    question_file = "data/mt_bench/question.jsonl"
+    questions = load_questions(question_file, question_begin, question_end)
+
+    if not answer_file:
```
Comment (alimaredia) [inline src/instructlab/eval/gen_api_answer.py:97]: This shouldn't hold up the merging of this PR, but it seems that if `gen_api_answer.run()` is executed multiple times the answer file would be overwritten. Would we want to introduce a flag that copies the answer file over somewhere and makes the filename somewhat unique?

File: data/mt_bench/reference_answer/gpt-4-turbo.jsonl [hunk 4]
```diff
@@ -0,0 +1,80 @@
...se landscapes from volcanic earth to lush rainforests offered adventures both thrilling and spiritual. At Hawaii Volcanoes National Park, awe struck me as I observed the powerful natural forces at work. Another profound experience awaited at Pu'uhonua o Honaunau National Historical Park, where ancient laws and traditions underscored the deep cultural reverence for peace and forgiveness.\n\nAs my journey concluded in Kauai, affectionately known as \"The Garden Isle,\" every moment was filled with natural splendor. Ascending the trails of Waimea Canyon, the \"Grand Canyon of the Pacific,\" amazed me with its vibrant spectrum of colors. Adventures along the Napali Coast, whether by boat or helicopter, afforded unparalleled views of dramatic cliffs and verdant valleys.\n\nAs reflections on my Hawaiian journey abound, appreciation for the hospitality and traditions of its people deepens. A visit to Hawaii offers more than picturesque scenery; it's an immersive experience where culture, history, and nature intertwine beautifully. Anticipate enchantment, arrive with an open heart, and allow the islands' magic to sweep you into an unforgettable adventure."]}], "tstamp": 1714190221.1310115}
```
Comment (alimaredia) [inline data/mt_bench/reference_answer/gpt-4-turbo.jsonl:1]: @xukai92 is this file needed at all? From what I remember we're only using `gpt-4.jsonl`.

File: steps_mt_bench.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+# MT-Bench Broken Down in Eval Scripts (not PR Bench)
```
Comment (alimaredia) [inline steps_mt_bench.md:1]: @danmcp will this file eventually get removed? Maybe we evolve it into the README for the repo?

File: src/instructlab/eval/mtbench.py (hunk 1 above)
Comment (danmcp) [reply src/instructlab/eval/mtbench.py:42]: Good question.  Maybe we can do some autotuning.

File: steps_mt_bench.md (hunk 5 above)
Comment (danmcp) [reply steps_mt_bench.md]: I merged with the README for now.  Still work to do obviously.

File: data/mt_bench/reference_answer/gpt-4-turbo.jsonl (hunk 4 above)
Comment (danmcp) [reply data/mt_bench/reference_answer/gpt-4-turbo.jsonl:1]: It's not needed for now.

File: src/instructlab/eval/gen_api_answer.py (hunk 3 above)
Comment (danmcp) [reply src/instructlab/eval/gen_api_answer.py:97]: We could use MTBenchEvaluator to generate an id and put all the results into an <id> folder between generate and judge.  Unclear what the desired experience is yet.

File: src/instructlab/eval/mtbench.py (hunk 1 above)
Comment (danmcp) [reply src/instructlab/eval/mtbench.py]: I went ahead an pulled the variable out to the top as an option (max_workers)

**Review by JamesKunstle — COMMENTED** [review]
First-round of review with a few comments/requests. Code is very well written albeit challenging to understand fully (a side-effect of pandas and file processing),

File: README.md [hunk 6]
```diff
@@ -5,4 +5,90 @@
...
+    └── rc
+        ├── model_answer
+        │   └── instructlab
+        │       └── granite-7b-lab.jsonl
+        ├── question.jsonl
+        └── reference_answer
+            └── gpt-4.jsonl
+```
+
+For running judge model with vllm make sure you run with `--served-model-name gpt-4`
```
[the remark this answers is not in the document]
Comment (JamesKunstle) [reply README.md:54]: i.e. vllm shouldn't be targetting gpt-4?

[the remark this answers is not in the document]
Comment (danmcp) [reply README.md:54]: It's temporary and in place because of the supported model list.  Can rework once the new judge is in place.

File: README.md [hunk 7]
```diff
@@ -5,4 +5,90 @@
...
+
+You should run with `--tensor-parallel-size <NUM GPUS>` and possibly increase `--max-model-len` to increase the context length
```
[the remark this answers is not in the document]
Comment (danmcp) [reply README.md:56]: Running vllm from here is temporary.  Will need to be launched from the cli.

File: src/instructlab/eval/common.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    """Load model answers.
+
+    The return value is a python dict of type:
+    Dict[model_name: str -> Dict[question_id: int -> answer: dict]]
+    """
+    filenames = glob.glob(os.path.join(answer_dir, "*.jsonl"))
+    filenames.sort()
+    model_answers = {}
+    for filename in filenames:
+        model_name = os.path.basename(filename)[:-6]
```
[the remark this answers is not in the document]
Comment (danmcp) [reply src/instructlab/eval/common.py:97]: Fixed, thanks!

File: src/instructlab/eval/judgment.py
[the remark this answers is not in the document]
Comment (danmcp) [reply src/instructlab/eval/judgment.py:1]: I tried to hit everything with at least a basic description

File: src/instructlab/eval/judgment.py [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+    judge,
+    ref_answers=None,
+    multi_turn=False,
+):
+    matches = []
+    for q in questions:
+        if multi_turn and len(q["turns"]) != 2:
+            continue
+        q_id = q["question_id"]
+        for _, m in enumerate(models):
```
[the remark this answers is not in the document]
Comment (danmcp) [reply src/instructlab/eval/judgment.py]: Fixed, thanks!

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (danmcp) [reply README.md]: I went ahead and fixed this so it will take any model you pass it as the judge

File: requirements.txt [hunk 10]
```diff
@@ -1 +1,11 @@
...
+FastChat
+GitPython>=3.1.42,<4.0.0
+shortuuid
+openai<1.0.0
+psutil
+torch
+transformers
+accelerate
+pandas
+pandas-stubs
```
Comment (nathan-weinberg) [inline requirements.txt:11]: @cdoern any requirements issues here compared to the CLI?

File: src/instructlab/eval/answers.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    output_base_dir = bench_dir(output_dir, bench_name, branch)
+
+    question_file = f"{data_base_dir}/question.jsonl"
+    questions = load_questions(question_file, question_begin, question_end)
+
+    answer_file = f"{output_base_dir}/model_answer/{model_name}.jsonl"
+    if os.path.isfile(answer_file):
+        os.remove(answer_file)
+    # print(f"Output to {answer_file}")
+
```
Comment (nathan-weinberg) [inline src/instructlab/eval/answers.py:117]: Should uncomment or remove - not opposed to keeping for debugging purposes (can always remove later)

File: src/instructlab/eval/common.py [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+API_ERROR_OUTPUT = "$ERROR$"
+
+# Categories that need reference answers
+NEED_REF_CATS = ["math", "reasoning", "coding", "arena-hard-200", "taxonomy"]
+
+# Extract scores from judgments
+one_score_pattern = re.compile(r"\[\[(\d+\.?\d*)\]\]")
+one_score_pattern_backup = re.compile(r"\[(\d+\.?\d*)\]")
+
+# Sampling temperature configs for
```
Comment (nathan-weinberg) [inline src/instructlab/eval/common.py:31]: Missing the end of the sentence?

File: src/instructlab/eval/common.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+    questions = questions[begin:end]
+    return questions
+
+
+def load_model_answers(answer_dir: str, model_name=None):
+    """Load model answers.
+
+    The return value is a python dict of type:
+    Dict[model_name: str -> Dict[question_id: int -> answer: dict]]
+    """
```
Comment (nathan-weinberg) [inline src/instructlab/eval/common.py:86]: Ret value can also be specified in type hint, e.g.

```python
def load_model_answers(answer_dir: str, model_name=None) -> dict:

File: src/instructlab/eval/common.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+                    break
+    return model_answers
+
+
+def load_judge_prompts(prompt_file: str):
+    """Load judge prompts.
+
+    The return value is a python dict of type:
+    Dict[judge_name: str -> dict]
+    """
```
Comment (nathan-weinberg) [inline src/instructlab/eval/common.py:110]: Same comment as above

```python
def load_judge_prompts(prompt_file: str) -> dict:

File: src/instructlab/eval/common.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+            "judgment": judgment,
+            "score": score,
+            "turn": turn,
+            "tstamp": time.time(),
+        }
+        # print(
+        #    f"question: {question_id}, turn: {turn}, model: {model}, "
+        #    f"score: {score}, "
+        #    f"judge: {(judge.model_name, judge.prompt_template['name'])}"
+        # )
```
Comment (nathan-weinberg) [inline src/instructlab/eval/common.py:210]: Should remove or keep for debugging (fine with keeping personally)

File: src/instructlab/eval/data/mt_bench/reference_answer/gpt-4.jsonl [hunk 16]
```diff
@@ -0,0 +1,30 @@
+{"question_id": 101, "answer_id": "TFomieEmmAgdeCkvmuvwbc", "model_id": "gpt-4", "choices": [{"index": 0, "turns": ["If you have just overtaken the second person, your current position is now second place. The person you just overtook is now in third place.", "If you have just overtaken the last person, it means you were previously the second to last person in the race. After overtaking the last person, your position remains the same, which is second to last. The person you just overtook is now in the last place."]}], "tstamp": 1686286924.844282}
```
Comment (nathan-weinberg) [inline src/instructlab/eval/data/mt_bench/reference_answer/gpt-4.jsonl:1]: Is it fine that `model_id` is `gpt-4` even though we are using Prometheus2 instead of GPT-4?

File: src/instructlab/eval/pr_bench_generator.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+        for file in files:
+            if file.split("/")[-1] == "qna.yaml":
+                file_paths.append(os.path.join(root, file))
+    return file_paths
+
+
+def read_qna(fn):
+    with open(fn, "r", encoding="utf-8") as file:
+        contents = yaml.safe_load(file)
+    return contents.get("seed_examples")
```
Comment (nathan-weinberg) [inline src/instructlab/eval/pr_bench_generator.py:28]: Are we checking the YAML/schema validity at all?

Don't think we need for this PR specifically, but @bjhargrave is working on a Dev Doc to get this functionality into the `instructlab-schema` package: https://github.com/instructlab/dev-docs/pull/101

**Review by nathan-weinberg — COMMENTED** [review]
Initial round of comments, will look more in-depth as well

File: src/instructlab/eval/data/mt_bench/reference_answer/gpt-4.jsonl (hunk 16 above)
Comment (danmcp) [reply src/instructlab/eval/data/mt_bench/reference_answer/gpt-4.jsonl:1]: Yes, the reference answers are provided by gpt-4, the model answers are provided by the model being judged.  And the judgment of how good the model being judged is compared to the gpt-4 answer is provided by prometheus.

File: src/instructlab/eval/answers.py (hunk 11 above)
Comment (danmcp) [reply src/instructlab/eval/answers.py]: Done

File: src/instructlab/eval/common.py (hunk 13 above)
Comment (danmcp) [reply src/instructlab/eval/common.py]: Done

File: src/instructlab/eval/common.py (hunk 15 above)
Comment (danmcp) [reply src/instructlab/eval/common.py]: Done

File: src/instructlab/eval/common.py (hunk 14 above)
Comment (danmcp) [reply src/instructlab/eval/common.py]: Done

File: README.md [hunk 18]
```diff
@@ -5,4 +5,103 @@
...
+pip install -r requirements-dev.txt
+pip install -e .
+pip install vllm==0.3.3
+python -m vllm.entrypoints.openai.api_server --model instructlab/granite-7b-lab --tensor-parallel-size 1
+```
+
+In another shell window
+
+```shell
+OPENAI_API_KEY="NO_API_KEY" python3 tests/test_gen_answers.py
```
Comment (alimaredia) [inline README.md:36]: @danmcp what do you think about just setting this environment variable in `gen_answers()` and `judge_answers()` in the `MTBenchEvaluator` class? This shouldn't block merging this PR though.

File: tests/test_branch_gen_answers.py [hunk 19]
```diff
@@ -0,0 +1,10 @@
+# First Party
+from instructlab.eval.mt_bench import MTBenchBranchEvaluator
+
+mt_bench_branch = MTBenchBranchEvaluator(
+    "instructlab/granite-7b-lab",
+    "instructlab/granite-7b-lab",
+    "../taxonomy",
```
Comment (alimaredia) [inline tests/test_branch_gen_answers.py:7]: @danmcp maybe I'm missing context or another PR that's being worked on, do these tests run in an automated fashion anywhere yet? 

The setup for this test would include cloning a taxonomy repo and then passing that path in right?

**Review by alimaredia — APPROVED** [review]

File: tests/test_branch_gen_answers.py (hunk 19 above)
Comment (danmcp) [reply tests/test_branch_gen_answers.py:7]: Not automated yet.  Instructions are in the README for now.

File: src/instructlab/eval/common.py (hunk 12 above)
Comment (danmcp) [reply src/instructlab/eval/common.py]: Fixed, thanks!

File: src/instructlab/eval/mt_bench_common.py [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+
+def chat_completion_openai(model, conv, temperature, max_tokens, api_dict=None) -> str:
+    if api_dict is not None:
+        openai.api_base = api_dict["api_base"]
+        openai.api_key = api_dict["api_key"]
+    output = API_ERROR_OUTPUT
+    for _ in range(API_MAX_RETRY):
+        try:
+            messages = conv.to_openai_api_messages()
+            if messages[0]["role"] == "system" and messages[1]["role"] == "user":
```
[the remark this answers is not in the document]
Comment (danmcp) [reply src/instructlab/eval/mt_bench_common.py]: I went ahead and included

**Review by JamesKunstle — APPROVED** [review]
This PR seems like it's in good enough shape to start iterating on.

**Review by nathan-weinberg — APPROVED** [review]
